src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    /// <summary>Provides a set of connection pools, each for its own endpoint.</summary>
+    internal sealed class HttpConnectionPools : IDisposable
+    {
+        /// <summary>How frequently an operation should be initiated to clean out old pools and connections in those pools.</summary>
+        private const int CleanPoolTimeoutMilliseconds =
+#if DEBUG
+            1_000;
+#else
+            30_000;
+#endif
+        /// <summary>The pools, indexed by endpoint.</summary>
+        private readonly ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool> _pools;
+        /// <summary>Timer used to initiate cleaning of the pools.</summary>
+        private readonly Timer _cleaningTimer;
+        /// <summary>The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</summary>
+        private readonly int _maxConnectionsPerServer;
+
+        /// <summary>Initializes the pools.</summary>
+        /// <param name="maxConnectionsPerServer">The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</param>
+        public HttpConnectionPools(int maxConnectionsPerServer)
+        {
+            _maxConnectionsPerServer = maxConnectionsPerServer;
+            _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
```

Comment:
Ideally, I think we would not have a cleanup timer unless there are actually active connections.  That is, if/when all connections are closed, we'd disable the timer, until a new connection is created.

I don't think this needs to be addressed in the current PR, but it's something to keep in mind.